src/Storages/MergeTree/IntegerCodecTrait.h

```diff
@@ -0,0 +1,232 @@
+#pragma once
+#if defined(__x86_64__) || defined(_M_X64)
+#if defined(__SSE4_1__)
+#define USE_SIMDCOMP 1
+#endif
+#endif
+
+extern "C"
+{
+#if defined(USE_SIMDCOMP)
+#include <simdcomp.h>
+#endif
+}
+
+namespace DB
+{
+
+/// Block codec used by PostingsContainerImpl to compress/decompress arrays of
+/// unsigned integers (typically delta/gap values).
+struct BlockCodec
+{
+    static constexpr size_t kBlockSize = 128;
+
+    /// Returns {compressed_bytes, bits} where bits is the max bit-width required
+    /// to represent all values in [0..n).
+    ALWAYS_INLINE static std::pair<size_t, size_t> evaluateSizeAndMaxBits([[maybe_unused]] const uint32_t * data, size_t n)
+    {
+#if defined(USE_SIMDCOMP)
+        auto bits = maxbits_length(data, n);
+        auto bytes = simdpack_compressedbytes(n, bits);
+        return {bytes, bits};
+#else
+        auto bits = maxBits(data, n);
+        auto bytes = static_cast<uint32_t>((n * static_cast<size_t>(bits) + 7) / 8);
+        return {bytes, bits};
+#endif
+    }
+
+    ALWAYS_INLINE static uint32_t encode(uint32_t *in, std::size_t n, [[maybe_unused]] uint32_t bits, unsigned char * & out)
+    {
+#if defined(USE_SIMDCOMP)
+        /// simdcomp expects __m128i* output pointer; we compute consumed bytes
+        /// from the returned end pointer (in units of 16-byte vectors).
+        auto *m128_out = reinterpret_cast<__m128i *>(out);
+        auto *m128_out_end = simdpack_length(in, n, m128_out, bits);
+        auto used = static_cast<size_t>(m128_out_end - m128_out) * sizeof(__m128i);
+        out += used;
+        return used;
+#else
+        if (n != kBlockSize)
+            return encodeVariadic(in, n, bits, out);
+        return encodeGeneric(in, bits, out);
+#endif
+    }
+
+    ALWAYS_INLINE static std::size_t decode(unsigned char * & in, std::size_t n, [[maybe_unused]] uint32_t bits, uint32_t * out)
+    {
+#if defined(USE_SIMDCOMP)
+        /// simdcomp expects __m128i* input pointer; we compute consumed bytes
+        /// from the returned end pointer (in units of 16-byte vectors).
+        auto *m128i_p = reinterpret_cast<__m128i *>(in);
```

Comment:
Before this cast should we validate that `n` is big enough? Or the API can manage this propperly?